File: solver.cpp
// -*- coding:utf-8 -*-

//
// Pentomino Puzzle Solver with C++ (c++17)
//

const char* PIECE_DEF_DOC =R"(
+-------+-------+-------+-------+-------+-------+
|       |   I   |  L    |  N    |       |       |
|   F F |   I   |  L    |  N    |  P P  | T T T |
| F F   |   I   |  L    |  N N  |  P P  |   T   |
|   F   |   I   |  L L  |    N  |  P    |   T   |
|       |   I   |       |       |       |       |
+-------+-------+-------+-------+-------+-------+
|       | V     | W     |   X   |    Y  | Z Z   |
| U   U | V     | W W   | X X X |  Y Y  |   Z   |
| U U U | V V V |   W W |   X   |    Y  |   Z Z |
|       |       |       |       |    Y  |       |
+-------+-------+-------+-------+-------+-------+
)";


#include <iostream>
#include <algorithm>
#include <vector>
#include <sstream>
#include <map>

namespace std {
  std::string to_string( const char &c ) { return "'" + string(1, c) + "'"; }
  std::string to_string( const std::string &s ) { return s; }

  template< typename C >
  std::string
  join( const C& v,
        const char* delim = nullptr,
        const char* bracket = nullptr ) {
    std::string str;
    if ( !v.empty() ) {
      if ( delim == nullptr )  delim = "";
      auto it = v.begin();
      str += to_string( *it );
      while ( ++it != v.end() ) { str += delim + to_string( *it ); }
    }
    return bracket? ( bracket[0] + str + bracket[1] ) : str;
  }

  template<typename C>
  std::string
  to_string( const C &v ) {
    return join( v, ", ", "[]" );
  }

}


std::vector<std::string>
split( const std::string &s, char delim, bool f = true )
{
  std::vector<std::string> elems;
  std::stringstream ss(s);
  std::string item;
  while (getline(ss, item, delim)) {
    if ( f || !item.empty()) {
      elems.push_back(item);
    }
  }
  return elems;
}

/////////////////////////////////////////////////////////////

struct Point
{
  int x, y;

  Point( int x_ = 0 , int y_ = 0 ) : x( x_ ), y( y_ )  { }
};

bool
operator==( const Point &a, const Point &b )
{
  return a.y == b.y && a.x == b.x;
}


std::string
to_string( const Point &pt )
{
  return to_string( std::vector<int>( {pt.x, pt.y} ) );
}

using Fig  = std::vector<Point>;

///////////////////////////////////////////////////////////////

std::map<char,Fig> make_piece_def()
{
  std::map<char,Fig> fig_def;
  Point xy{ 0, 0 };
  for ( auto c : std::string( PIECE_DEF_DOC ) ) {
    if ( std::isalpha( c ) )  fig_def[ c ].push_back( { xy.x / 2, xy.y } );
    xy = ( c == '\n' )? Point{ 0, xy.y + 1 } : Point{ xy.x + 1, xy.y };
  }
  return fig_def;
}

std::map<char,Fig> PIECE_DEF = make_piece_def();

///////////////////////////////////////////////////////////////

bool debug_flg = false;

class Piece
{
public:
  std::vector<Fig> figs;
  char    id;
  Piece*  next = nullptr;

  Piece( char id_, const Fig& def_, Piece *next_ )
    : id ( id_ ), next( next_ )
  {
    for ( auto r_f = 0; r_f < 8; ++r_f ) {   // rotate and flip
      Fig fig = def_;            // copy

      for ( auto &p : fig ) {    // p: reference
        for ( auto n = 0; n < r_f % 4; ++n ) { p = { -p.y, p.x }; } // rotate
        if ( r_f >= 4 )                      { p = { -p.x, p.y }; } // flip
      }

      std::sort( fig.begin(), fig.end(),                            // sort
                 []( const auto&a, const auto&b ) {
                   return (a.y != b.y)? (a.y < b.y) : (a.x < b.x);
                 });

      for ( auto it = fig.rbegin(); it != fig.rend(); ++it ) {      // normalize
        it->x -= fig[0].x;
        it->y -= fig[0].y;
      }

      if ( std::find( figs.begin(), figs.end(), fig ) == figs.end() ) { // uniq
        figs.push_back( fig );
      }
    }

    if ( debug_flg ) {
      std::cout << id << " : (" << figs.size() << ")" << std::endl;
      for ( auto &fig : figs ) {
        std::cout << "    " << std::to_string( fig ) << std::endl;
      }
    }
  }

};  // class Piece

//         2
// (-1,-1) | (0,-1)
//   ---4--+--1----
// (-1, 0) | (0, 0)
//         8
const std::vector<std::string> ELEMS[2] = {
  //     0      3     5    6    7     9    10   11   12   13   14   15
  split("    ,,,+---,,----,+   ,+---,,+---,|   ,+---,+   ,+---,+   ,+---", ','),
  split("    ,,,    ,,    ,    ,    ,,|   ,|   ,|   ,|   ,|   ,|   ,|   ", ',')
};

class Board
{
public:
  static constexpr char SPACE = ' ';

  int   width, height;
  std::vector<std::vector<char>> cells;

  Board( int w, int h )
  {
    width  = w;
    height = h;
    cells.assign( height, std::vector<char>( width, SPACE ) );
    if ( w * h == 64 ) {    // 8x8 or 4x16
      cells[ h/2 - 1 ][ w/2 - 1 ] = '@';  cells[ h/2 - 1 ] [ w/2 ] = '@';
      cells[ h/2     ][ w/2 - 1 ] = '@';  cells[ h/2     ] [ w/2 ] = '@';
    }
  }

  char
  at( int x, int y )
  {
    return (x >= 0 && x < width && y >= 0 && y < height)? cells[ y ][ x ] : '?';
  }


  bool
  check( const Point &o, const Fig &fig )
  {
    for ( const auto &pt : fig ) {
      if ( at( o.x + pt.x, o.y + pt.y ) != SPACE ) return false;
    }
    return true;
  }


  void
  place( const Point &o, const Fig &fig, char id )
  {
    for ( const auto &pt : fig ) {
      cells[ o.y + pt.y ][ o.x + pt.x ] = id;
    }
  }


  Point
  find_space( const Point &xy  )
  {
    auto x = xy.x;
    auto y = xy.y;

    while ( cells[ y ][ x ] != SPACE ) {
      if ( ++x == width )  { x = 0;  ++y; }
    }

    return Point( x, y );
  }


  std::string
  render()
  {
    std::vector<std::string> lines;

    for ( auto y = 0; y <= height; ++y ) {
      for ( auto d = 0; d < 2; ++d ) {
        lines.push_back( "" );
        for ( auto x = 0; x <= width; ++x ) {
          int code =
            ( ( at( x+0, y+0 ) != at( x+0, y-1 ) )? 1 : 0 ) |
            ( ( at( x+0, y-1 ) != at( x-1, y-1 ) )? 2 : 0 ) |
            ( ( at( x-1, y-1 ) != at( x-1, y+0 ) )? 4 : 0 ) |
            ( ( at( x-1, y+0 ) != at( x+0, y+0 ) )? 8 : 0 );
          lines.back() += ELEMS[ d ][ code ];
        }
      }
    }

    return join( lines, "\n" );
  }

}; // class Board


class Solver
{
  int    solutions;
  Board  &board;
  Piece  *Unused;

public:
  Solver( int width, int height ) :
    solutions( 0 ), board( *new Board(width, height) )
  {
    Piece *pc = nullptr;
    std::string ids( "FLINPTUVWXYZ" );   // the first 'X' is dummy
    for ( auto it = ids.rbegin(); it != ids.rend(); ++it ) {
      pc = new Piece( *it, PIECE_DEF[ *it ], pc );
    }
    Unused = new Piece( '!', Fig{}, pc );         // dumy piece
    // limit the symmetry of 'F'
    pc->figs.resize( (width == height)? 1 : 2);
  }


  void
  solve( const Point& xy_ )
  {
    if ( Unused->next != nullptr ) {
      auto xy = board.find_space( xy_ );
      Piece *prev = Unused;
      while ( Piece *pc = prev->next ) {
        prev->next = pc->next;
        for ( const auto &fig : pc->figs ) {
          if ( board.check( xy, fig ) ) {
            board.place( xy, fig, pc->id );
            solve( xy );
            board.place( xy, fig, Board::SPACE );
          }
        }
        prev = ( prev->next = pc );
      }
    }
    else {
      ++solutions;

      std::string curs_up = "";
      if ( solutions > 1 ) {
        curs_up = "\033[" + std::to_string( board.height * 2 + 2 ) + "A";
      }
      std::cout << curs_up << board.render() << solutions << std::endl;
    }
  }

};  // class Solver

int
main( int argc, char *argv[] )
{
  int width = 6,  height = 10;
  for ( auto i = 1; i < argc; ++i ) {
    if ( std::string( argv[i] ) == "--debug" ) debug_flg = true;
    if ( std::string( argv[i] ) == "-d"      ) debug_flg = true;
    int  w = 0, h = 0;
    char c = 0;
    std::istringstream( argv[i] ) >> w >> c >> h;
    if ( w >= 3 && h >= 3 && ( w * h == 60 || w * h == 64 ) ) {
      width  = w;
      height = h;
    }
  }

  Solver solver( width, height );
  solver.solve( Point( 0, 0 ) );

  return 0;
}

File: solver.pl
# -*- coding:utf-8 -*-

#
# Pentomino Puzzle Solver with Perl
#

$PIECE_DEF=<<EOT;
+-------+-------+-------+-------+-------+-------+
|       |   I   |  L    |  N    |       |       |
|   F F |   I   |  L    |  N    |  P P  | T T T |
| F F   |   I   |  L    |  N N  |  P P  |   T   |
|   F   |   I   |  L L  |    N  |  P    |   T   |
|       |   I   |       |       |       |       |
+-------+-------+-------+-------+-------+-------+
|       | V     | W     |   X   |    Y  | Z Z   |
| U   U | V     | W W   | X X X |  Y Y  |   Z   |
| U U U | V V V |   W W |   X   |    Y  |   Z Z |
|       |       |       |       |    Y  |       |
+-------+-------+-------+-------+-------+-------+
EOT

sub piece_def {
    my ($id ) = @_;

    my ( $x, $y ) = ( 0, 0 );
    my @def = ();
    foreach my $ch ( split( "", $PIECE_DEF ) ) {
        if ( $ch =~ $id  ) { push( @def, [ int($x/2), $y ] );  }
        if ( $ch =~ /\n/ ) { $y++;  $x = 0; }
        else               { $x++;          }
    }
    return \@def;
}

sub fig_to_str {
    my ( $fig ) = @_;
    return "[ " .
        join( " ", map( sprintf( "(%d,%d)", $_->[0], $?->[1] ), @$fig ) ) .
        " ]";
}

##############################################
$debug_flg  = 0;

sub new_Piece {
    my ($id ) = @_;

    my $def = piece_def( $id );

    my $figs_str = "";
    my @figs = ();
    for ( my $r_f = 0; $r_f < 8; $r_f++ ) {     # rotate & flip
        my @fig = ();

        foreach my $pt_ ( @$def ) {
            my @pt = ( $pt_->[0], $pt_->[1] );    # copy
            for ( my $r = 0; $r < $r_f % 4; $r++ ) {            # rotate
                @pt = ( -$pt[1], $pt[0] );
            }
            @pt = ( -$pt[0], $pt[1] )  if ( $r_f >= 4 );        # flip
            push( @fig, \@pt );
        }

        @fig = sort{                                            # sort
            ($a->[1] == $b->[1])? ($a->[0] <=> $b->[0]) : ($a->[1] <=> $b->[1])
        }( @fig );

        my ($ox, $oy) = ($fig[0][0], $fig[0][1] );              # normalize
        @fig = map( [ $_->[0] - $ox, $_->[1] - $oy ], @fig );

        if ( index( $figs_str, fig_to_str( \@fig ) ) == -1 ) {   # uniq
            $figs_str = $figs_str . fig_to_str( \@fig );
            push( @figs, \@fig );
        }
    }

    if ( $debug_flg ) {
        printf( "%s: (%d)\n", $id, scalar(@figs) );
        foreach my $fig ( @figs ) {
            printf( "\t%s\n",  fig_to_str( $fig ) );
        }
    }
    return \@figs;
}

##############################################

sub new_Board {
    my ( $w, $h ) = @_;
    $width  = $w;
    $height = $h;

    use constant SPACE => " ";
    @cells = ();
    for ( my $y = 0; $y < $height; $y++ ) {
        for ( my $x = 0; $x < $width; $x++ ) {
            $cells[$y][$x] = SPACE;
        }
    }
    if ( $width * $height == 64 ) {     # 8x8 or 4x16
        place( $width/2-1, $height/2-1, [ [0,0], [0,1], [1,0], [1,1] ], "@" );
    }

    my @elems = (
        "    ,,,+---,,----,+   ,+---,,+---,|   ,+---,+   ,+---,+   ,+---",
        "    ,,,    ,,    ,    ,    ,,|   ,|   ,|   ,|   ,|   ,|   ,|   "
        );
    @ELEMS = map( [ split( /,/, $_ ) ], @elems );
}


sub at {
    my ( $x, $y ) = @_;
    if ( $x >= 0 && $x < $width && $y >= 0 && $y < $height ) {
        return $cells[$y][$x];
    }
    return "?";
}


sub check {
    my ( $x, $y, $fig ) = @_;
    foreach $pt ( @{$fig} ) {
        return 0  if at( $pt->[0] + $x, $pt->[1] + $y ) ne SPACE;
    }
    return 1;
 }


sub place {
    my ( $x, $y, $fig, $id ) = @_;
    foreach $pt ( @{$fig} ) {
        $cells[ $pt->[1] + $y ][ $pt->[0] + $x ] = $id;
    }
 }


sub find_space {
    my ( $x, $y ) = @_;
    while ( $cells[ $y ][ $x ] ne SPACE ) {
        ( $x = ( $x + 1 ) % $width ) == 0 && $y++;
    }
    return [ $x, $y ];   # reference
}


sub render {
    my @lines = ();
    for ( my $y = 0; $y <= $height; $y++ ) {
        for ( my $d = 0; $d < 2; $d++ ) {
            my $line = "";
            for ( $x = 0; $x <= $width; $x++ ) {
                my $code =
                    ( at( $x+0, $y+0 ) ne at( $x+0, $y-1 )? 1 : 0 ) |
                    ( at( $x+0, $y-1 ) ne at( $x-1, $y-1 )? 2 : 0 ) |
                    ( at( $x-1, $y-1 ) ne at( $x-1, $y+0 )? 4 : 0 ) |
                    ( at( $x-1, $y+0 ) ne at( $x+0, $y+0 )? 8 : 0 );
                $line = $line . $ELEMS[ $d ][ $code ];
            }
            push( @lines, $line );
        }
    }
    return join( "\n", @lines );
}

##############################################

sub new_Solver {
    my ( $width, $height ) = @_;

    %Pieces    = ();
    %Next      = ();
    $Solutions = 0;

    my $pc = "";
    foreach my $id ( reverse( split( "", "FILNPTUVWXYZ" ) ) ) {
        $Pieces{ $id } = new_Piece( $id );
        $Next{ $id }   = $pc;
        $pc = $id;
    }
    # limit the symmetry of 'F'
    $Pieces{ "F" } = [@{$Pieces{ "F" }}[0.. (($width==$height)? 0 : 1 )]];
    $Unused = "!";                  # dummy piece
    $Next{ $Unused } = "F";
    new_Board( $width, $height );
}


sub solve {
    my ( $x_, $y_ ) = @_;
    if ( $Next{ $Unused } ne "" ) {
        my ( $x, $y ) = @{find_space( $x_, $y_ )};
        my $prev      = $Unused;
        while ( ( my $pc = $Next{ $prev } ) ne "" ) {
            $Next{ $prev } = $Next{ $pc };
            foreach my $fig ( @{$Pieces{ $pc }} ) {
                if ( check( $x, $y, $fig ) ) {
                    place( $x, $y, $fig, $pc );
                    solve( $x, $y );
                    place( $x, $y, $fig, SPACE );
                }
            }
            $prev = ( $Next{ $prev } = $pc );
        }
    }
    else {
        $Solutions ++;
        my $curs_up = ($Solutions > 1)? sprintf( "\e[%dA", $height * 2 + 2 ) : "";
        printf( "%s%s%d\n", $curs_up, render(), $Solutions );
        #my $line = <STDIN>;
    }
}

#################################

my ( $width, $height ) = @{[ 6, 10 ]};

foreach my $arg ( @ARGV ) {
    if ( $arg eq "--debug" ) {
        $debug_flg = 1;
    }
    my @sz = $arg =~ /^(\d+)\D(\d+)$/;
    @sz = map( int($_), @sz );
    if ( scalar(@sz) == 2 && $sz[0] >= 3 && $sz[1] >= 3 &&
         ( $sz[0] * $sz[1] == 60 || $sz[0] * $sz[1] == 64 ) ) {
        ( $width, $height ) = @sz;
    }
}

new_Solver( $width, $height );
solve( 0, 0 );
